runtime/src/accounts_background_service.rs

```diff
@@ -0,0 +1,178 @@
+                    &self.snapshot_config.compression,
+                );
+                if r.is_err() {
+                    warn!(
+                        "Error generating snapshot for bank: {}, err: {:?}",
+                        snapshot_root_bank.slot(),
+                        r
+                    );
+                }
+                snapshot_time.stop();
+
+                // Cleanup outdated snapshots
+                let mut purge_old_snapshots_time = Measure::start("purge_old_snapshots_time");
+                snapshot_utils::purge_old_snapshots(&self.snapshot_config.snapshot_path);
+                purge_old_snapshots_time.stop();
+
+                datapoint_info!(
+                    "handle_snapshot_requests-timing",
+                    ("shrink_time", shrink_time.as_us(), i64),
+                    ("clean_time", clean_time.as_us(), i64),
+                    ("snapshot_time", snapshot_time.as_us(), i64),
+                    (
+                        "purge_old_snapshots_time",
+                        purge_old_snapshots_time.as_us(),
+                        i64
+                    )
+                );
+                snapshot_root_bank.block_height()
+            })
+    }
+}
+
+pub struct AccountsBackgroundService {
+    t_background: JoinHandle<()>,
+}
+
+impl AccountsBackgroundService {
+    pub fn new(
+        bank_forks: Arc<RwLock<BankForks>>,
+        exit: &Arc<AtomicBool>,
+        snapshot_request_handler: Option<SnapshotRequestHandler>,
+    ) -> Self {
+        info!("AccountsBackgroundService active");
+        let exit = exit.clone();
+        let mut consumed_budget = 0;
+        let mut last_cleaned_block_height = 0;
+        let t_background = Builder::new()
+            .name("solana-accounts-background".to_string())
+            .spawn(move || loop {
+                if exit.load(Ordering::Relaxed) {
+                    break;
+                }
+
+                // Grab the current root bank
+                let bank = bank_forks.read().unwrap().root_bank().clone();
+
+                // Check to see if there were any requests for snapshotting banks
+                // < the current root bank `bank` above.
+
+                // Claim: Any snapshot request for slot `N` found here means that cleanup has
+                // been called at most on slot `N`
+                //
+                // Proof: Assume for contradiction that we find a snapshot request for slot `N` here,
+                // but cleanup has already happened on some slot `M > N`.
+                //
+                // Then that means on some previous iteration of this loop, we found `bank_forks.root_bank() == M`,
+                // but did not see the snapshot for `N`. However, this is impossible because BankForks.set_root()
+                // will always flush the snapshot request for `N` to the snapshot request channel
+                // before setting a root `M > N`.
```

Comment:
I think this clam still holds true but I see different proof more suited here.

Proof: The updating of bank_forks.root_bank() to new `root` and flushing new snapshot request into `snapshot_request_sender` (if any) are done atomically at `bank_forks::set_root()` while both being locked with the same `RwLock`. 

So, it's not possible to observe inconsistent state of root slot and snapshot request bank slot.

Reply:
> // ... this is impossible because BankForks.set_root()
> // will always flush the snapshot request for `N` to the snapshot request channel
> // before setting a root `M > N`.

In fact, I think [settting root](https://github.com/solana-labs/solana/blob/f0e02d25881b9d421359d56740a659acbc49ef0e/runtime/src/bank_forks.rs#L190) happens before [flushing a snapshot request](https://github.com/solana-labs/solana/pull/12194/files?diff=split&w=1#diff-59ca0c8d9d4c1a3d9080b5609feba39dR228)

Reply:
> flushing a snapshot request

So the proof is trying to show that if we got a snapshot request for `S`, we cannot have called clean on a slot `R > S` in some previous iteration of the loop.

> In fact, I think setting root happens before flushing a snapshot request.
Yeah agreed the setting of the root `R` happens happens before the flushing, but what that statement is saying is that all snapshots requests `S<R` should have been flushed already, so by here: https://github.com/solana-labs/solana/pull/12194/files?diff=split&w=1#diff-649a1adaf14e9c6dcde42a5b36ff2e47R146-R149, if we find a snapshot request for slot `S`, then the last time we called `bank.clean_accounts(true)` in the previous iteration of the loop must have been for `R <= S`. Updated the proof to hopefully make that clearer!